cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1305,6 +1313,245 @@ def _check_column_availability(self):
+                "'EFO:0010961' (Visium Spatial Gene Expression) and uns['spatial']['is_single'] is True."
+            )
+            # Exit as library_id is missing.
+            return
+
+        # Confirm shape of library_id is valid: allowed keys are images and scalefactors.
+        library_id_key = library_ids[0]
+        uns_library_id = uns_spatial[library_id_key]
+        if not self._has_no_extra_keys(uns_library_id, ["images", "scalefactors"]):
+            self.errors.append(
+                "uns['spatial'][library_id] can only contain the keys 'images' and 'scalefactors'."
+                f"Detected keys: {list(uns_library_id.keys())}."
+            )
+
+        # images is required.
+        if "images" not in uns_library_id:
+            self.errors.append("uns['spatial'][library_id] must contain the key 'images'.")
+        # images is specified: proceed with validation of images.
+        else:
+            # Confirm shape of images is valid: allowed keys are fullres and hires.
+            uns_images = uns_library_id["images"]
+            if not self._has_no_extra_keys(uns_images, ["fullres", "hires"]):
+                self.errors.append(
+                    "uns['spatial'][library_id]['images'] can only contain the keys 'fullres' and 'hires'."
+                    f"Detected keys: {list(uns_images.keys())}."
+                )
+
+            # hires is required.
+            if "hires" not in uns_images:
+                self.errors.append("uns['spatial'][library_id]['images'] must contain the key 'hires'.")
+            # hires is specified: proceed with validation of hires.
+            else:
+                self._validate_spatial_image_shape("hires", uns_images["hires"], 2000)
+
+            # fullres is optional.
+            uns_fullres = uns_images.get("fullres")
+            if uns_fullres is None:
+                # Warn if no fullres is specified as it is strongly recommended.
+                self.warnings.append(
+                    "No uns['spatial'][library_id]['images']['fullres'] was found. "
+                    "It is STRONGLY RECOMMENDED that uns['spatial'][library_id]['images']['fullres'] is provided."
+                )
+            else:
+                self._validate_spatial_image_shape("fullres", uns_fullres)
+
+        # scalefactors is required.
+        if "scalefactors" not in uns_library_id:
+            self.errors.append("uns['spatial'][library_id] must contain the key 'scalefactors'.")
+        # scalefactors is specified: proceed with validation of scalefactors.
+        else:
+            # Confirm shape of scalefactors is valid: allowed keys are spot_diameter_fullres and tissue_hires_scalef.
+            uns_scalefactors = uns_library_id["scalefactors"]
+            if not self._has_no_extra_keys(uns_scalefactors, ["spot_diameter_fullres", "tissue_hires_scalef"]):
```

Comment:
If I set `uns['spatial'][library_id][scalefactors]` = anything other than a `dict` or `pandas.Dataframe`, I get an AttributeError: "AttributeError: '{type}' object has no attribute 'keys'" This behaves similarly to my above comment where the raised exception prematurely exits validation.

`uns['spatial'][library_id][scalefactors]` is an [output](https://scanpy.readthedocs.io/en/latest/generated/scanpy.read_visium.html) from `scanpy.read_visium()` which would suggest typing should be pretty consistent and not the wide range I've tested against. 

But we are finding through the curation towards 5.1.0 that images that are labeled as a given resolution aren't necessarily the correct resolution or set at the proper scale factor. So we do have to manually adjust these keys and values and we could easily wrangle to something other than a dict with correct keys and values.